doc/man1/roarlight.1
.\" roarlight.1:

.TH "roarlight" "1" "January 2010" "RoarAudio" "System User's Manual: roarlight"

.SH NAME

roarlight \- controls the Light Control subsystem of RoarAudio

.SH SYNOPSIS

roarlight [OPTIONS]... command [command...]

.SH DESCRIPTION

This command works simular to \fBroarctl\fR but controls actual data on the
Light Control subsystem.

.SH "OPTIONS"

.TP
\fB--server    SERVER\fR
Set server hostname

.TP
\fB--help\fR
Show this help

.SH "COMMANDS"

.TP
\fBhelp\fR
Show this help

.TP
\fBsleep TIME\fR
Sleeps for TIME seconds

.TP
\fBset   chan=val\fR
Set a DMX Channel

.SH "SEE ALSO"
\fBroarctl\fR(1),
\fBroartips\fR(7),
\fBlibroar\fR(7),
\fBRoarAudio\fR(7).

.SH "HISTORY"

For history information see \fBRoarAudio\fR(7).

.\" ll


ChangeLog
v. 0.3beta3 - ?
	* added some aRts/NAS/YIFF/PulseAudio tools to compbins
	* added support for comp libs to roar-config
	* added support for uni-directional streams in roarinterconnect
	* write much better support for config, including codec config
	* reduced the TCP overhead produced by speex
	* added support for speex preprocessor
	* get names of complibs from table, do not guess them
	* a lot small fixes
	* moved most of the mixer stuff out of roard into libroardsp
	* added flags immutable and enhance
	* test for signal.h
	* Corrected (DSTR) HTTP Protocol support so servers without headerflush
	  should now work, too.
	* added protocol support for icy
	* Added version numbers to lib names
	* updated manpages
	* wrote OSS emulation

v. 0.3beta2 - Mon Oct 05 2009 19:48 CEST
	* Added prethru flag
	* Added support to read info about client's network node
	* Ported libroaresd, libroarartsc
	* Fixed some problems with EsounD emulation
	* added support for compatibility binarys
	* added poll() support to libroarsndio
	* reject streams of unsupported codec

v. 0.3beta1 - Sat Sep 26 2009 25:50 CEST
	* incremented nummber of streams per client fron 4 to 8
	* RIFF WAVE codecfilter uses virtual streams
	* another set of changed names of parameters in roarctl
	* Interface to read byteorder of a client from roard
	* implemented a 'state of stream' var
	* cleanup, small fixes
	* added VIO interface to transcoder interface
	* ask codecfilter of parent in case of new child stream
	* added 'subsystem' complex: multi subsystem data on single streams
	* added basic /Radio Data and Trasmitter Control System/.
	* better support for stereo upmixing
	* make audio group a optional thing
	* better support for win32
	* wrote WMM driver
	* better multi protocol support
	* Support for server side EsounD emulation on win32

v. 0.3beta0 - Sun Sep 06 2009 22:42 CEST
	* added new stream directions for MIDI and light control
	* cleanup of client stream struct
	* added basic light control subsystem
	* added DMX512 driver
	* wrote basic MIDI subsystem
	* added flag 'mute'
	* added flag 'mmap'
	* added lib libroarmidi and libroarlight
	* added OpenSLP support
	* rewote fillmixbuer2 code
	* added a roar_conv2() with a more clear API
	* added support to write pidfile
	* added support for encoding 32 bit Vorbis streams
	* added transcoding features in libroardsp
	* added tool 'roarphone'
	* added tool 'roarinterconnect'
	* added simple raw data subsystem
	* added working support for THRU streams
	* changed names of parameters in roarctl
	* added lib libroareio - Extended IO for RoarAudio
	* Update order of default drivers, now it is: oss, ao, esd, null
	* cleanup of old sysio driver interface
	* added server side EsounD emulation
	* added multi homed support to server
	* fixed a long outstanding bug with pa_path_get_filename()

v. 0.2 - Thu May 21 2009 21:00
	* do a some more testings for functions not existing on win32
	* init winsock
	* done a lot fixes in order to get libroar and some clients work on win32
	* done a lot cleanup on configure script

v. 0.2rc2 - Sun May 17 2009 25:44
	* corrected check for GNU peercred
	* added basic logging support to configure
	* fixed plugin configures
	* added and fixed semi-install target on plugins
	* call plugin configures from main configure
	* include plugins in Makefile known to work from configure
	* fixed Makefiles of plugins: added -I../../include, -L../../lib
	* Make use of pkg-config in plugin configures
	* corrected possition of -ls on linker commands in most Makefiles
	* added more explicide linker commands so most things now build
	  with classic BSD make.
	* devided into components: libs, roard, clients, doc
	* add toption to set roard's cycle freq
	* moved a lot things in plugin's configure scripts to a commone one

v. 0.2rc1 - Wed May 13 2009 12:05
	* Started to port to OpenSolaris
	* fixed bug in default based sockets, make HTTP VIO from non default port work
	* got libroarsndio build under OpenBSD 4.5
	* do not depend tests in configure on roaraudio.h: fixes some checks on *BSD
	* renamed member 'fh' of connection struct to '__fh', this breaks
	  all code depending on it so we see how it behaves in case
	  we replace this with a real VIO object in near future
	* started VIO 2popen interface
	* fixed exec() on systems with sizeof(int) != sizeof(char*)
	* do not do unnessesery calls to fdatasync() on sockets
	* link roarvorbis against libogg and libvorbis
	* remove need of GNU/Tar like -z on remote system from send-roaraudio
	* added support for hwmixer flag on OSS and sndio driver
	* fiexed some typos

v. 0.2beta3 - Fri May 01 2009 15:15
	* added esdfilt options to roarfilt, esdplay options to roarcatplay
	* added short options to roarcat/mon/filt as stated in the manpage
	* updated manpages
	* also build libroar{,dsp}.a
	* Changed VIO CTLs in a binary inkompatible way
	* added support for autoconf flag to OSS driver
	* added driver for OpenBSD sndio interface
	* started libroarsndio
	* added code for plugins: audacirus, mplayer, gst10
	* fixed some building problems
	* added roarify and s start for it's manpage
	* added stream flags pause and hwmixer

v. 0.2beta2 - Fri Mar 27 2009 21:20
	* added (dummy) flags autoconf and cleanmeta
	* added streams_ctl() to server
	* added support for primary flag as -oO, support for blocks parameter
	* added (GnuPG) pinentry support.
	* added support for Ogg Vorbis encoder to set meta data :)
	* fixed some (minor) bugs, including segfault in vio
	* added a lot things to the vio interface
	* added gzip and simple OpenPGP support to vio interface
	* added a simple pointer stack.
	* added (OpenSSH) askpass support.
	* added vio pipe support, currectly pipe() and socketpair() based onse.
	* test for OpenSSL
	* added support to VIO open OpenSSL BIOs
	* added support to read a string and convert it into a VIO stack
	* added simple VIO stack as VIO pass object
	* started cleaner VIO Socket implementation

roarclients/roarlight.c
//roarlight.c:

/*
 *      Copyright (C) Philipp 'ph3-der-loewe' Schafft - 2009
 *
 *  This file is part of roarclients a part of RoarAudio,
 *  a cross-platform sound system for both, home and professional use.
 *  See README for details.
 *
 *  This file is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License version 3
 *  as published by the Free Software Foundation.
 *
 *  RoarAudio is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this software; see the file COPYING.  If not, write to
 *  the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
 *
 */

#include <roaraudio.h>
#include <libroarlight/libroarlight.h>

#define CONVAR struct roar_connection * con

void usage (void) {
 printf("roarlight [OPTIONS]... command [command...]\n");

 printf("\nOptions:\n\n");

 printf("  --server    SERVER    - Set server hostname\n"
        "  --help                - Show this help\n"
       );

 printf("\nCommands:\n\n");
 printf(
        "  help                    - Show this help\n"
        "  sleep TIME              - Sleeps for TIME seconds\n"
        "  set   chan=val          - Set a DMX Channel\n"
       );
}

int cmd_set (CONVAR, char * arg) {
 char * next = arg;
 char * k, * v;
 int32_t chan, val;
 struct roar_roardmx_message mes;
 int fh;                    //TODO: we should use pure VIO here
 struct roar_vio_calls vio;

 roar_roardmx_message_new_sset(&mes);

 while (next != NULL) {
  arg  = next;
  next = strstr(next, ",");
  if ( next != NULL ) {
   *next = 0;
    next++;
  }

  k = arg;
  v = strstr(arg, "=");
  if ( v == NULL )
   return -1;

  *v = 0;
   v++;

  chan = atoi(k);
  val  = atoi(v);
//  printf("k='%s'(%i), v='%s'(%i)\n", k, chan, v, val);
  if ( roar_roardmx_message_add_chanval(&mes, chan, val) == -1 )
   return -1;
 }

 if ( (fh = roar_simple_new_stream(con,
                                   ROAR_RATE_DEFAULT, ROAR_CHANNELS_DEFAULT, ROAR_BITS_DEFAULT,
                                   ROAR_CODEC_ROARDMX, ROAR_DIR_LIGHT_IN
                                  )) == -1 )
  return -1;

 if ( roar_vio_open_fh_socket(&vio, fh) == -1 ) {
  close(fh);
  return -1;
 }

 if ( roar_roardmx_message_send(&mes, &vio) == -1 ) {
  roar_vio_close(&vio);
  return -1;
 }

 roar_vio_close(&vio);

 return 0;
}

int main (int argc, char * argv[]) {
 char * server   = NULL;
 char * k;
 int    i;
 struct roar_connection con;

 for (i = 1; i < argc; i++) {
  k = argv[i];

  if ( !strcmp(k, "--server") || !strcmp(k, "-s") ) {
   server = argv[++i];
  } else if ( !strcmp(k, "--codec") ) {

  } else if ( !strcmp(k, "--help") || !strcmp(k, "-h") ) {
   usage();
   return 0;
  } else if ( *k == '-' ) {
   fprintf(stderr, "Error: unknown argument: %s\n", k);
   usage();
   return 1;
  } else {
   break;
  }
 }

 if ( roar_connect(&con, server) == -1 ) {
  fprintf(stderr, "Error: Can not connect to server\n");
  return 1;
 }

 if ( roar_identify(&con, "roarlight") == -1 ) {
  fprintf(stderr, "Error: Can not identify to server\n");
  roar_disconnect(&con);
  return 1;
 }

 if ( i == argc ) {
  fprintf(stderr, "Error: No Commands given\n");
  return 0; // this is not a fatal error...
 }

 for (; i < argc; i++) {
  k = argv[i];
  // cmd is in k

  printf("--- [ %s ] ---\n", k);

  if ( !strcmp(k, "help") ) {
   usage();

  } else if ( !strcmp(k, "sleep") ) {
   sleep(atoi(argv[++i]));

  } else if ( !strcmp(k, "set") ) {
   i++;
   if ( cmd_set(&con, argv[i]) == -1 ) {
    fprintf(stderr, "Error: can not set channels\n");
   } else {
    printf("channels changed\n");
   }

  } else {
   fprintf(stderr, "Error: invalid command: %s\n", k);
  }
 }

 roar_disconnect(&con);

 return 0;
}

//ll
